### PR TITLE
[GUVNOR-3339] Implement tests which scale Kie server to zero in Opens…

### PR DESCRIPTION
**Pull request by jakubschwan**
…hift

### CODE REVIEW

File: test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/survival/KieServerWithWorkbenchSurvivalIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+        kieServerControllerClientProvider = new KieServerControllerClientProvider(workbenchWithKieServerScenario.getWorkbenchDeployment());
+        kieServerMgmtControllerClient = kieServerControllerClientProvider.getKieServerMgmtControllerClient();
+        kieServerControllerClientProvider.waitForServerTemplateCreation();
+    }
+
+    @After
+    public void tearDown() {
+        InstanceLogUtil.writeDeploymentLogs(workbenchWithKieServerScenario.getWorkbenchDeployment());
+        InstanceLogUtil.writeDeploymentLogs(workbenchWithKieServerScenario.getKieServerDeployment());
+        workbenchWithKieServerScenario.undeploy();
```
Comment (sutaakar) [inline test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/survival/KieServerWithWorkbenchSurvivalIntegrationTest.java]: Please add here if (.. != null) check to prevent NPE when for example gitProvider is null. For example see https://github.com/kiegroup/kie-cloud-tests/pull/22

File: test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/survival/KieServerWithWorkbenchSurvivalIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+        logger.debug("Start process instance");
+        Long signalPid = kieServerClientProvider.getProcessClient().startProcess(CONTAINER_ID, SIGNALTASK_PROCESS_ID);
+        Assertions.assertThat(signalPid).isNotNull().isGreaterThan(0L);
+        Assertions.assertThat(kieServerClientProvider.getProcessClient().findProcessInstances(CONTAINER_ID, 0, 10)).isNotNull().hasSize(1);
+
+        scaleKieServerTo(0, serverInfo.getServerId());
+
+        logger.debug("Start process instance");
+        try {
+            Long noContainerPid = kieServerClientProvider.getProcessClient().startProcess(CONTAINER_ID, SIGNALTASK_PROCESS_ID);
```
Comment (sutaakar) [inline test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/survival/KieServerWithWorkbenchSurvivalIntegrationTest.java]: This try/catch approach isn't ideal. It is not clear what should be the result and what do you verify in that case. Try to use Assertions.assertThatThrownBy() instead.

File: test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/survival/KieServerWithWorkbenchSurvivalIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+        scaleKieServerTo(1, serverInfo.getServerId());
+
+        logger.debug("Register Kie Container to Kie Server");
+        kieServerMgmtControllerClient.saveContainerSpec(serverInfo.getServerId(), serverInfo.getName(), CONTAINER_ID, CONTAINER_ALIAS, PROJECT_GROUP_ID, PROJECT_NAME, PROJECT_VERSION, KieContainerStatus.STARTED);
+        kieServerClientProvider.waitForContainerStart(CONTAINER_ID);
+
+        logger.debug("Start process instance");
+        Long signalPid = kieServerClientProvider.getProcessClient().startProcess(CONTAINER_ID, SIGNALTASK_PROCESS_ID);
+        Assertions.assertThat(signalPid).isNotNull().isGreaterThan(0L);
```
Comment (sutaakar) [inline test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/survival/KieServerWithWorkbenchSurvivalIntegrationTest.java]: You can add Assertions as static import.

File: test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/survival/KieServerWithWorkbenchSurvivalIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        } catch (KieServicesHttpException ex) {
+            // ok - response code 403 Forbidden
+            // The application is currently not serving requests at this endpoint. It may not have been started or is still starting.
+            logger.debug("KieServicesHttpException catched:\n" + ex.toString());
+        }
+
+        scaleKieServerTo(1, serverInfo.getServerId());
+
+        logger.debug("Wait for Kie Container start again");
+        kieServerClientProvider.waitForContainerStart(CONTAINER_ID);
```
Comment (sutaakar) [inline test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/survival/KieServerWithWorkbenchSurvivalIntegrationTest.java]: Not needed. Managed Kie server is ready when it installs all containers defined in server template. This is checked in previous step - scaleKieServerTo.

File: test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/survival/KieServerWithWorkbenchSurvivalIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+        logger.debug("Start new process instance");
+        Long newPid = kieServerClientProvider.getProcessClient().startProcess(CONTAINER_ID, SIGNALTASK_PROCESS_ID);
+        Assertions.assertThat(newPid).isNotNull().isGreaterThan(0L);
+
+        logger.debug("Check strated processes");
```
Comment (sutaakar) [inline test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/survival/KieServerWithWorkbenchSurvivalIntegrationTest.java]: strated - > started

File: test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/survival/KieServerWithWorkbenchSurvivalIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+        Assertions.assertThat(kieServerClientProvider.getProcessClient().findProcessInstances(CONTAINER_ID, 0, 10)).isNotNull().hasSize(2);
+
+        ProcessInstance processInstance = kieServerClientProvider.getProcessClient().getProcessInstance(CONTAINER_ID, signalPid);
+        Assertions.assertThat(processInstance).isNotNull();
+        Assertions.assertThat(processInstance.getState()).isNotNull().isEqualTo(org.kie.api.runtime.process.ProcessInstance.STATE_ACTIVE);
+
+        kieServerClientProvider.getProcessClient().abortProcessInstance(CONTAINER_ID, signalPid);
```
Comment (sutaakar) [inline test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/survival/KieServerWithWorkbenchSurvivalIntegrationTest.java]: No need for process abort. The internal DB is deleted after this test.

File: test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/survival/KieServerWithWorkbenchSurvivalIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+        ProcessInstance processInstance = processServicesClient.getProcessInstance(CONTAINER_ID, signalPid);
+        assertThat(processInstance).isNotNull();
+        assertThat(processInstance.getState()).isNotNull().isEqualTo(org.kie.api.runtime.process.ProcessInstance.STATE_ACTIVE);
+    }
+
+    private void scaleKieServerTo(int count, String serverId) {
+        logger.debug("Scale Kie Server to " + count);
+        deploymentScenario.getKieServerDeployment().scale(count);
+        deploymentScenario.getKieServerDeployment().waitForScale();
+        checkServerTemplateInstanceCount(serverId, count);
```
Comment (sutaakar) [inline test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/survival/KieServerWithWorkbenchSurvivalIntegrationTest.java]: I would put this "checkServerTemplateInstanceCount" method directly into the code - to be more transparent about test behaviour. If such check is placed here then method name "scaleKieServerTo" has misleading name as it doesn't tell that it contains also check.

File: test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/survival/KieServerWithWorkbenchSurvivalIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+        logger.debug("Start process instance");
+        Long signalPid = processServicesClient.startProcess(CONTAINER_ID, SIGNALTASK_PROCESS_ID);
+        assertThat(signalPid).isNotNull().isGreaterThan(0L);
+        assertThat(queryServicesClient.findProcessInstances(0, 10)).isNotNull().hasSize(1);
+
+        scaleKieServerTo(0, serverInfo.getServerId());
+
+        logger.debug("Start process instance");
+
+        Assertions.assertThatThrownBy(() -> {
```
Comment (sutaakar) [inline test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/survival/KieServerWithWorkbenchSurvivalIntegrationTest.java]: Please use static import.

File: test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/survival/KieServerWithWorkbenchSurvivalIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+            processServicesClient.startProcess(CONTAINER_ID, SIGNALTASK_PROCESS_ID);
+        }).isInstanceOf(KieServicesHttpException.class).hasMessageContaining("The host doesn't exist.");
```
Comment (sutaakar) [inline test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/survival/KieServerWithWorkbenchSurvivalIntegrationTest.java]: Please check HTTP code. I think it should return HTTP 503.

File: test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/survival/KieServerWithWorkbenchSurvivalIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+//        } catch (KieServicesHttpException ex) {
+//            // ok - response code 503 Forbidden
+//            // The application is currently not serving requests at this endpoint. It may not have been started or is still starting.
+//            System.out.println("KieServicesHttpException catched:\n" + ex.toString());
+//        }
+
+        scaleKieServerTo(1, serverInfo.getServerId());
+
+        logger.debug("Wait for Kie Container start again");
+        KieServerClientProvider.waitForContainerStart(deploymentScenario.getKieServerDeployment(), CONTAINER_ID);
```
Comment (sutaakar) [inline test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/survival/KieServerWithWorkbenchSurvivalIntegrationTest.java]: No need to wait for container to start. When Kie server is scaled then it should contain all containers.

File: test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/survival/KieServerWithWorkbenchSurvivalIntegrationTest.java [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+    protected WorkbenchWithKieServerScenario createDeploymentScenario(DeploymentScenarioBuilderFactory deploymentScenarioFactory) {
+        return deploymentScenarioFactory.getWorkbenchWithKieServerScenarioBuilder().build();
+    }
+
+    @Before
+    public void setUp() {
+        WorkbenchUtils.deployProjectToWorkbench(gitProvider, deploymentScenario.getWorkbenchDeployment(), PROJECT_NAME);
+
+        kieServerMgmtControllerClient = KieServerControllerClientProvider.getKieServerMgmtControllerClient(deploymentScenario.getWorkbenchDeployment());
+        KieServerControllerClientProvider.waitForServerTemplateCreation(deploymentScenario.getWorkbenchDeployment());
```
Comment (sutaakar) [inline test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/survival/KieServerWithWorkbenchSurvivalIntegrationTest.java]: This isn't needed. This synchronization happens already when deploying a scenario.

File: test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/survival/KieServerWithWorkbenchSurvivalIntegrationTest.java [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+                .register(new Authenticator(deploymentScenario.getKieServerDeployment().getUsername(),
+                                deploymentScenario.getKieServerDeployment().getPassword()))
+                .build();
+
+        WebTarget target = httpKieServerClient.target(deploymentScenario.getKieServerDeployment().getUrl().toString() + "/services/rest/server/containers");
+        Response response = target.request().get();
+        response.close();
+        assertThat(response.getStatus()).isEqualTo(Response.Status.SERVICE_UNAVAILABLE.getStatusCode());
+
+        URL postUrl = null;
```
Comment (sutaakar) [inline test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/survival/KieServerWithWorkbenchSurvivalIntegrationTest.java]: This part with POST request isn't needed as the previous GET call checks that Kie server is unavailable.

File: test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/survival/KieServerWithWorkbenchSurvivalIntegrationTest.java [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+import org.kie.server.integrationtests.controller.client.KieServerMgmtControllerClient;
+import org.kie.server.integrationtests.shared.filter.Authenticator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class KieServerWithWorkbenchSurvivalIntegrationTest extends AbstractCloudIntegrationTest<WorkbenchWithKieServerScenario> {
+
+    private static final String PROJECT_GROUP_ID = "org.kie.server.testing";
```
Comment (jiripetrlik) [inline test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/survival/KieServerWithWorkbenchSurvivalIntegrationTest.java]: We use this particular GAV quite often. Maybe it would be good to have these constants somewhere globally accessible.

File: test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/survival/KieServerWithWorkbenchSurvivalIntegrationTest.java [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+
+    private void sendRESTCallsToUnavailableKieServer() {
+        Client httpKieServerClient = new ResteasyClientBuilder()
+                .establishConnectionTimeout(10, TimeUnit.SECONDS)
+                .socketTimeout(10, TimeUnit.SECONDS)
+                .register(new Authenticator(deploymentScenario.getKieServerDeployment().getUsername(),
+                                deploymentScenario.getKieServerDeployment().getPassword()))
+                .build();
+
+        WebTarget target = httpKieServerClient.target(deploymentScenario.getKieServerDeployment().getUrl().toString() + "/services/rest/server/containers");
```
Comment (jiripetrlik) [inline test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/survival/KieServerWithWorkbenchSurvivalIntegrationTest.java]: Make a constant from "/services/rest/server/containers".

File: test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/survival/KieServerWithWorkbenchSurvivalIntegrationTest.java (hunk 13 above)
Comment (jakubschwan) [inline test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/survival/KieServerWithWorkbenchSurvivalIntegrationTest.java]: Good point, but this should be change in all test classes. I'll create a new PR with changes.

Comment (jiripetrlik) [inline test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/survival/KieServerWithWorkbenchSurvivalIntegrationTest.java]: Looks good now.

**Comment by jiripetrlik** [comment]
Looks good now. Merged.

File: test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/survival/KieServerWithWorkbenchSurvivalIntegrationTest.java (hunk 13 above)
Comment (jiripetrlik) [inline test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/survival/KieServerWithWorkbenchSurvivalIntegrationTest.java]: Yes, you are right. It will be better to do it in a separate PR. Maybe we can create constants for all GAVs which we use in tests.
